libdnf/base/transaction.cpp

```diff
@@ -495,22 +490,17 @@ void Transaction::Impl::set_transaction(rpm::solv::GoalPrivate & solved_goal, Go
         rpm::Package new_package(base, rpm::PackageId(id));
         auto reason = new_package.get_reason();
         TransactionPackage tspkg(new_package, TransactionPackage::Action::REINSTALL, reason);
-        tspkg.replaces.emplace(rpm::Package(base, rpm::PackageId(obs[0])));
-        for (int i = 1; i < obs.size(); ++i) {
-            rpm::Package obsoleted(base, rpm::PackageId(obs[i]));
+        for (auto replaced_id : obs) {
+            rpm::Package obsoleted(base, rpm::PackageId(replaced_id));
```

Comment:
This code is awfully similar between the various actions, though there are slight differences, like the variable above is called `obsoleted` here and `replaced_pkg` in the downgrades block above. I'm not entirely sure but it appears this should be factored out into a function to deduplicate and unify the code.

Reply:
Correct, Upgrade, reinstall, and downgrade have the same code. I will think about it.

Reply:
Fixed